### ISSUE TITLE
ExchangeDeclare ignores type argumen after exception

### ISSUE THREAD

**Issue by pardahlman** [rabbitmq/rabbitmq-dotnet-client#163]
I'm implementing an extension for `RawRabbit` to "re-declare" topology features in RabbitMq (https://github.com/pardahlman/RawRabbit/issues/75). I noticed an unexpected behaviour, which can be exemplified with this code

```csharp
using(var connection = new ConnectionFactory().CreateConnection())
using (var channel = connection.CreateModel())
{
	channel.ExchangeDeclare("example_exchange", ExchangeType.Topic);
	try
	{
		channel.ExchangeDeclare("example_exchange", ExchangeType.Direct);
	}
	catch (Exception) { /* do nothing */ }
	channel.ExchangeDeclare("example_exchange", ExchangeType.Topic); //  exception here
}
```
When running this code I get the following exception

> The AMQP operation was interrupted: AMQP close-reason, initiated by Peer, code=406, text="PRECONDITION_FAILED - inequivalent arg 'type' for exchange 'example_exchange' in vhost '/': **received 'direct' but current is 'topic'**", classId=40, methodId=10, cause=

The problem is that I'm not trying to declare the exchange as a `Direct`, but as `Topic`. The problem does not appear if I remove the try/catch blocks.

**Comment by michaelklishin** [comment]
2nd invocation should fail since the arguments are not equivalent but channel error takes time to reach the client.

Please post questions to rabbitmq-users in the future.

**Comment by pardahlman** [comment]
The same exception appears if I, instead of trying to re-declare the exchange, try to delete it (`channel.ExchangeDelete("example_exchange")`)

**Comment by michaelklishin** [comment]
@pardahlman on what line? Like I said, error delivery is asynchronous and lines number 1 and 2 is asking for this specific channel exception.

**Comment by michaelklishin** [comment]
Also, not sure if you've captured a protocol trace with [Wireshark](http://www.rabbitmq.com/amqp-wireshark.html) it it would make much easier for you to reason about the sequence of methods that are sent by both client and server.

**Comment by pardahlman** [comment]
@michaelklishin are you sure? I've added some additional work and `Thread.Sleep` but still get the same result

```csharp
using(var connection = new ConnectionFactory().CreateConnection())
using (var channel = connection.CreateModel())
using (var secondChannel = connection.CreateModel())
{
	channel.ExchangeDeclare("example_exchange", ExchangeType.Topic);
	try
	{
		channel.ExchangeDeclare("example_exchange", ExchangeType.Direct);

	}
	catch (Exception e)
	{
		 // the thrown exception is the expected OperationInterruptedException
	}

	secondChannel.ExchangeDeclare("first_exchange", ExchangeType.Direct); // this works
	secondChannel.ExchangeDeclare("second_exchange", ExchangeType.Direct); // this works
	secondChannel.ExchangeDeclare("third_exchange", ExchangeType.Direct); // this works
	secondChannel.ExchangeDeclare("forth_exchange", ExchangeType.Direct); // this works
	for (var i = 0; i < 100; i++) { Thread.Sleep(20); } //sleep for 2 seconds
	channel.ExchangeDelete("example_exchange"); // still throws
}
```

**Comment by michaelklishin** [comment]
@pardahlman I'm not sure what to recommend. Your test intentionally causes a channel exception, plus swallows an exception. It also uses multiple channels. The first non-swallowing operation throws an exception (by the time the channel should be closed).

The right thing to do is to correlate this with server logs and a Wireshark trace to see what's going on on the wire. What are you trying to accomplish with this test?

**Comment by michaelklishin** [comment]
I'm not sure what the 2nd channel and different exchange names is supposed to demonstrate here.

**Comment by pardahlman** [comment]
@michaelklishin maybe I misunderstood you when you said

>  channel error takes time to reach the client

I thought you meant that it would take time for the exception to reach the client, so I added a second channel to do some work and added thread sleeps etc to give the exception time to reach the client.

I'm reusing a channel in the application. One of the tasks that I want to do with the channel is to declare an exchange. Sometimes the exchange is already declared and perhaps with a different configuration that the one provided in the `ExchangeDeclare` statement. I want to catch that exception and act on that. After the exception is thrown I want to be able to re-use the channel. The first `ExchangeDeclare` was added so that it was easier to follow the example, I could have written the fist snippet this way 

```csharp
using(var connection = new ConnectionFactory().CreateConnection())
using (var channel = connection.CreateModel())
{
	try
	{
		channel.ExchangeDeclare("example_exchange", ExchangeType.Direct);
	}
	catch (OperationInterruptedException e)
	{
		 // exchange exists with different config
	}

	channel.ExchangeDelete("example_exchange"); // throws
}
```
This is perhaps expected behaviour, i.e. the channel can't be used once an exception has been thrown on it? Or if I re-phrase it as a question: given that I have a channel that has thrown an `OperationInterruptedException`, what should I do to be able to re-use it for other operations?
